feat: zeebe should write the requestid in the exported events/records

mustafadagher opened the pull request:
## Description
Zeebe should write the requestid in the exported events/records
- Moved `requestId` to `Record` interface
- Added requestId to zeebe-record-template
- Populated requestId in processingResultBuilder to have the same requestId appended in follow-up commands
- move RequestIdGenerator instantiation to be a broker startup step
   - this way the request id generation is not an implementation detail anymore, rather an explicit part of the engine
   - we can then ensure we have only one id generator per broker and hence the requestId is unique per cluster

## Related issues

closes #17637

mustafadagher committed:
feat: move requestId to Record interface

mustafadagher committed:
feat: add requestId to zeebe-record-template

mustafadagher committed:
feat: propagate requestId to subsequent commands

populate requestId in processingResultBuilder to have the same requestId appended in subsequent commands

mustafadagher committed:
test: should have `requestId` in the serialized Record json

mustafadagher committed:
feat: pass `requestId` to ProcessingResultBuilder also in case of error or rejection

mustafadagher commented:
> @Zelldon reminded me that we should verify that we do not make wrong assumptions about `requestId` being set. For example, maybe we assume somewhere that records with a `requestId` are user commands? That'd be wrong now.

That's a good point. 
❓  How/Where do you anticipate this assumption to be made? 
❓  And how would we avoid it? is communication somewhere (like in some docs (are there any to be updated?)) enough? or does it need code changes to make it somehow clearer?

mustafadagher commented:
I'm trying to look for possible ways to test/verify the changes here. But I couldn't come up with good ones yet. I'd appreciate any hints.

oleschoenburg commented:
>  ❓ How/Where do you anticipate this assumption to be made?

In our code only.

> ❓ And how would we avoid it? is communication somewhere (like in some docs (are there any to be updated?)) enough? or does it need code changes to make it somehow clearer?

I think it's not part of our public API so we'd only have to check internally. Just search for all usages of `requestId`. A bit tedious but I don't know how else we could do it.

berkaycanbc commented on tasklist/importer-850/src/main/java/io/camunda/tasklist/zeebeimport/v850/record/RecordImpl.java:
```diff
@@ -48,6 +48,8 @@ public class RecordImpl<T extends RecordValue> implements Record<T> {
 
   private Map<String, Object> authorizations;
 
+  private long requestId;
```
❓Looks fine to me but do we need someone from Tasklist team to review this part?

mustafadagher replied:
@houssain-barouni would this be something you might review?

mustafadagher replied:
Thanks @marcosgvieira ! 🙇‍♂️

berkaycanbc commented on zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingStateMachine.java:
```diff
@@ -361,6 +361,8 @@ private void batchProcessing(final TypedRecord<?> initialCommand) {
 
       final var command = pendingCommands.removeFirst();
 
+      processingResultBuilder.setRequestId(command.getRequestId());
```
🔧 We set `requestId` here for each follow up command. Although, it should be enough if we set it only once from the initial command.

I couldn't validate my assumption. Would be good to verify with tests and make the change if it works.

mustafadagher replied:
If you notice `TypedEventWriter.appendFollowUpEvent()` is called from places where mostly there is no access to the `requestId`, and then later in the stacktrace when the `ProcessingResultBuilder.append()` was called to append those events, the `metadata` was constructed without a `requestId` as well (since there is no access to get it).
Hence, I needed to rely on the `ProcessingResultBuilder` instantiated in the beginning of processing (in the `ProcessingStateMachine`) to keep the original requestId in its state so it can be appended to the followup requests.

I'm open to have a call about this to explore different possibilities if you like. It looks ugly, and like a non-intuitive workaround, but I tried to look into several ways around it, and so far I failed.

berkaycanbc replied:
Alright, I overlook that we built metadata for followup events and commands as a new object. I thought, we pass the metadata along from the initial command.

berkaycanbc commented on zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingStateMachine.java:
```diff
@@ -361,6 +361,8 @@ private void batchProcessing(final TypedRecord<?> initialCommand) {
 
       final var command = pendingCommands.removeFirst();
 
+      processingResultBuilder.setRequestId(command.getRequestId());
```
❌ Setting `requestId` is an exceptional case here and non-trivial. Let's add a comment on top of it and mention the intention that we would like to set the initial `requestId` to followup commands as well inside `ProcessingResultBuilder` implementations and eventually export it for each record.

That way, other developers can reason about the decision and maintain the related code easily in the future.

berkaycanbc reviewed (changes requested):
Very nice job @mustafadagher, thanks! 🚀 

I added a few comments to improve the implementation, please take a look. 👀 

❌ Let's add a test to verify if the `requestId` is carried between **multiple** follow up commands. I think, `StreamProcessortTest` should be the proper class to test the behaviour.

oleschoenburg commented:
@deepthidevaki mentioned to me that the `requestId` is not necessarily unique across partitions. @mustafadagher Can you evaluate whether that breaks the intended use in Operate?

Zelldon commented:
> @deepthidevaki mentioned to me that the `requestId` is not necessarily unique across partitions. @mustafadagher Can you evaluate whether that breaks the intended use in Operate?

That is true for that we have the streamid

deepthidevaki commented:
It might be unique across partitions, although this was not the intention. Please check [here](https://github.com/camunda/zeebe/blob/main/zeebe/transport/src/main/java/io/camunda/zeebe/transport/impl/AtomixServerTransport.java#L45). But it is easy to break this assumption if we run this code per partition. Right now it is per broker. If we are relying on this guarantee, we should take measures to ensure this is not accidentally broken.

mustafadagher commented:
> It might be unique across partitions, although this was not the intention. Please check [here](https://github.com/camunda/zeebe/blob/main/zeebe/transport/src/main/java/io/camunda/zeebe/transport/impl/AtomixServerTransport.java#L45). But it is easy to break this assumption if we run this code per partition. Right now it is per broker. If we are relying on this guarantee, we should take measures to ensure this is not accidentally broken.

Thanks @deepthidevaki @Zelldon @oleschoenburg for bringing this to my attention! It's a very important hint.

I see indeed for the current design it should be unique still across nodes, since the nodeId is passed to the generator and it occupies a few bits from the generated sequence.

As @deepthidevaki mentioned indeed this only relies on the guarantee of having this run only on the broker. 

To avoid this breaking the assumptions from Operate's perspective, we could do one of two solutions:
1. refrain from using this requestId totally, and revert back to the original solution by asking operate to provide this reference in the request.
2. to include the partitionId/requestStreamId in the exported records and the client responses too. As a matter of fact, `partitionId` is already part of the exported records, so it would be up to us to also pass it back in the client responses, and inform operate to rely on both values together for their queries. Not sure if this would impact the performance on their side by any means (@sdorokhova could maybe share her thoughts.) 

For now, I'm leaning towards solution number 2. But I appreciate hearing your thoughts further, and also letting me know if there are any extra points I've missed or wrong assumptions I could have made.

Thanks a lot!

mustafadagher commented:
> Very nice job @mustafadagher, thanks! 🚀
> 
> I added a few comments to improve the implementation, please take a look. 👀
> 
> ❌ Let's add a test to verify if the `requestId` is carried between **multiple** follow up commands. I think, `StreamProcessortTest` should be the proper class to test the behaviour.

Thanks a lot @berkaycanbc for the comments.
All `StreamProcessorTest`s mock the underlying logic. I tried to add something there, but none of the existing tests for example will pass by the line where I set the `requestId` in the `ProcessingResultBuilder`. I'll give it another try with fresh eyes tomorrow.

korthout commented on zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/api/ProcessingResultBuilder.java:
```diff
@@ -87,4 +87,6 @@ ProcessingResultBuilder withResponse(
   ProcessingResult build();
 
   boolean canWriteEventOfLength(int eventLength);
+
+  void setRequestId(long requestId);
```
❌ It doesn't make sense to me that this method exists on the `ProcessingResultBuilder` interface. The interface is used by the engine to provide results, but this method is providing details from the stream platform that are not adjusted by the engine. 

Where used in `ProcessingStateMachine`, the code always first creates the concrete implementation `BufferedProcessingResultBuilder` before calling the `setRequestId` method on it. So we can easily remove it from the interface by using `var` rather than the interface name explicitly.

mustafadagher replied:
so you're suggesting to move it to the concrete implementation instead? I don't mind that.
When I added it to the interface, I also thought I might need to have it mocked somewhere in tests. but so far there's no need.

korthout reviewed (commented):
Nice work @mustafadagher 

👍 The change makes sense to me. I don't think the engine depends on the idea that a `requestId` means that it's a user command, but please verify this as previously suggested. 

💭 If we ever need to identify user commands specifically, we can use the combination of `requestId` and no `sourceRecordPosition`. When set, the `sourceRecordPosition` identifies the position of the command whose processing resulted in writing this record, i.e., a follow-up record. So, if no `sourceRecordPosition` is set **and** a `requestId` is set, we know it's a new command that a user appended.

Please have a look at my change request and suggestion.

mustafadagher committed:
test: verify requestId is populated to followup events

mustafadagher committed:
refactor: move `setRequestId` from the `ProcessingResultBuilder` to the concrete implementation

mustafadagher committed:
refactor: propagating the requestId from the initial command is enough

- moved setting the requestId to the top outside the loop
- renamed `setRequestId` to `withRequestId` to match the builder pattern

mustafadagher commented on zeebe/stream-platform/src/test/java/io/camunda/zeebe/stream/impl/StreamProcessorTest.java:
```diff
@@ -407,10 +409,18 @@ public void shouldProcessFollowUpEventsAndCommands() {
+      assertThat(followupMetadata.getRequestId()).isEqualTo(requestId);
     }
   }
 
```
@Zelldon just out of curiosity, asking since you were the author.
Is there any validation that the below test `shouldSetSourcePointerForFollowUpRecords` asserts and not covered by this above one (`shouldProcessFollowUpEventsAndCommands `)?

Zelldon replied:
I think the purpose/scope of both tests were different. But yeah based on the diff they look quite similar 🤔 I haven't check it in detail (and have no time for it yet) We could reduce the assertions in the first one, to reduce the overlap if you like and concentrate on the purpose of verifying that follow up commands are processed as well, I guess.

Zelldon replied:
Well, yeah I guess this a controversial topic I'm not sure whether I can discuss this here right now :D There are several use cases and benefits of repeating yourself in the matter of test cases. Sometimes you have to sacrifice one against another. For example, this allows us to highlight better where the issue lies and makes sure that one tests does one thing (SRP) as with one method which does many assertsion/verifications it is often hard to tell.

Zelldon replied:
But as I don't see currently an easy way of how to separate this, I'm find if you delete it.

mustafadagher replied:
I removed it, and enhanced the asserts with descriptions above! Thank you 🙇‍♂️

korthout commented on zeebe/stream-platform/src/test/java/io/camunda/zeebe/stream/util/RecordToWrite.java:
```diff
@@ -55,6 +55,11 @@ public static RecordToWrite command() {
     return new RecordToWrite(recordMetadata.recordType(RecordType.COMMAND));
   }
 
+  public static RecordToWrite command(final long requestId) {
+    final RecordMetadata recordMetadata = new RecordMetadata();
+    return new RecordToWrite(recordMetadata.recordType(RecordType.COMMAND).requestId(requestId));
+  }
```
😍 This is a great addition!

korthout reviewed (approved):
I like it! 🚀 

LGTM 👍

berkaycanbc commented on zeebe/stream-platform/src/test/java/io/camunda/zeebe/stream/impl/StreamProcessorTest.java:
```diff
@@ -407,10 +409,18 @@ public void shouldProcessFollowUpEventsAndCommands() {
     while (logStreamReader.hasNext()) {
-      assertThat(logStreamReader.next().getSourceEventPosition()).isEqualTo(firstRecordPosition);
+      final LoggedEvent followup = logStreamReader.next();
+      assertThat(followup.getSourceEventPosition()).isEqualTo(firstRecordPosition);
```
👍 Nice one. This exactly verifies if we are able to propagate the request id to follow up events. It mocks the record processor but not the processing result builder where the logic lies.

mustafadagher committed:
test: delete `shouldSetSourcePointerForFollowUpRecords` test case

All the asserts are already covered by `shouldProcessFollowUpEventsAndCommands`

berkaycanbc commented on zeebe/protocol/src/main/java/io/camunda/zeebe/protocol/record/Record.java:
```diff
@@ -129,6 +129,9 @@ public interface Record<T extends RecordValue> extends JsonSerializable {
   T getValue();
 
+  /**
+   * @return the id of the request that produced this record
```
🔧 Let's enrich this explanation as @korthout asked https://github.com/camunda/zeebe/pull/17877#discussion_r1586311562. For example, we can mention the request id is unique within the broker. Therefore, the investigation you did to find out its uniqueness will buy us time later on.

berkaycanbc reviewed (approved):
Thanks for additional changes @mustafadagher! ⭐ 

🔧 I asked for a non-blocking change related to information we have about `requestId` uniqueness.

⚠️ Since now we know that `requestId` is only unique within the broker, we should align this information with Operate team that they will not query records only with `requestId`. I saw you already communicated this with them here https://github.com/camunda/zeebe/issues/16916#issuecomment-2090077639. Let's do not merge this PR until we got response for your comment.

✅ Pre-approving it as the implementation on our side looks sufficient.

mustafadagher committed:
refactor: move RequestIdGenerator instantiation to be a broker startup step

- this way the request id generation is not an implementation detail anymore, rather an explicit part of the engine
- we can then ensure we have only one id generator per broker and hence the requestId is unique per cluster

korthout commented:
@mustafadagher With https://github.com/camunda/camunda/pull/18669 merged, is there anything from this PR that we want to keep?

mustafadagher commented:
> @mustafadagher With #18669 merged, is there anything from this PR that we want to keep?

@korthout, I wanted to open that up with @npepinpe, but didn't get the chance last week. Will try to find time for this the coming days
